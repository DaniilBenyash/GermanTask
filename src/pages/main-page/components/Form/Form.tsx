import { FC } from "react";
import styles from "./Form.module.scss";
import { Contact } from "../../api";
import { SubmitHandler, useForm } from "react-hook-form";
import { Button, Input, Form as FormUI } from "../../../../ui-kit";

type FormProps = {
  contact: Partial<Contact> | null;
  onAddUser: (contact: Contact) => void;
  onUpdateUser: (contact: Contact) => void;
  onDeleteUser: (id: number) => void;
};

export const Form: FC<FormProps> = ({ contact, onAddUser, onDeleteUser, onUpdateUser }) => {
  const {
    register,
    handleSubmit,
    reset,
    formState: { isValid },
  } = useForm<Contact>({
    defaultValues: { ...contact },
  });
  const onSubmit: SubmitHandler<Contact> = (contact) => {
    if (!contact.id) {
      onAddUser(contact);
      return;
    }
    onUpdateUser(contact);
  };

  const handleDelete = () => {
    if (!contact?.id) return;
    onDeleteUser(contact.id);
  };
  return (
    <FormUI onSubmit={handleSubmit(onSubmit)}>
      <div className={styles.data_container}>
        <h2 className={styles.title}>Eintrag bearbeiten</h2>
        <div className={styles.input_container}>
          <Input placeholder="Vorname*" {...register("firstname", { required: true })} />
          <Input placeholder="Nachname*" {...register("lastname", { required: true })} />
          <Input placeholder="E-Mail*" {...register("email", { required: true })} />
        </div>
      </div>
      <div className={styles.button_container}>
        {contact?.id && (
          <Button type="secondary" onClick={handleDelete}>
            LÃ–SCHEN
          </Button>
        )}
        <div className={styles.right_section_buttons}>
          <Button type="primary">ABBRECHEN</Button>
          <Button type="default" disabled={!isValid}>
            SPEICHERN
          </Button>
        </div>
      </div>
    </FormUI>
  );
};
