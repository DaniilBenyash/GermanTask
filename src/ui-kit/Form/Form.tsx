import { FC, ReactNode } from "react";
import styles from "./Form.module.scss";

type FormProps = {
  onSubmit: () => void;
  children: ReactNode;
};
export const Form: FC<FormProps> = ({ onSubmit, children }) => {
  return (
    <form className={styles.form} onSubmit={onSubmit}>
      {children}
    </form>
  );
};
