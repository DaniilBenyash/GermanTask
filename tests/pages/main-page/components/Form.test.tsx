import { Contact } from "../../../../src";
import { Form } from "../../../../src/pages/main-page/components/Form";
import renderer from "react-test-renderer";

it("renders correctly", () => {
  const tree = renderer
    .create(
      <Form
        contact={null}
        onAddUser={function (contact: Contact): void {
          throw new Error("Function not implemented.");
        }}
        onUpdateUser={function (contact: Contact): void {
          throw new Error("Function not implemented.");
        }}
        onDeleteUser={function (id: number): void {
          throw new Error("Function not implemented.");
        }}
        onCloseModal={function (): void {
          throw new Error("Function not implemented.");
        }}
      />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

// import { render, fireEvent, screen } from "@testing-library/react";
// import userEvent from "@testing-library/user-event";

// describe("Form Component Tests", () => {
//   let mockOnAddUser: jest.Mock;
//   let mockOnUpdateUser: jest.Mock;
//   let mockOnDeleteUser: jest.Mock;
//   let mockOnCloseModal: jest.Mock;

//   beforeEach(() => {
//     mockOnAddUser = jest.fn();
//     mockOnUpdateUser = jest.fn();
//     mockOnDeleteUser = jest.fn();
//     mockOnCloseModal = jest.fn();
//   });

//   it("renders correctly", () => {
//     const tree = renderer
//       .create(
//         <Form
//           contact={null}
//           onAddUser={mockOnAddUser}
//           onUpdateUser={mockOnUpdateUser}
//           onDeleteUser={mockOnDeleteUser}
//           onCloseModal={mockOnCloseModal}
//         />,
//       )
//       .toJSON();
//     expect(tree).toMatchSnapshot();
//   });

//   it("calls onAddUser when onSubmit is called with a new contact", () => {
//     render(
//       <Form
//         contact={null}
//         onAddUser={mockOnAddUser}
//         onUpdateUser={mockOnUpdateUser}
//         onDeleteUser={mockOnDeleteUser}
//         onCloseModal={mockOnCloseModal}
//       />,
//     );
//     // Assuming 'SPEICHERN' is the text on the submit button
//     userEvent.type(screen.getByPlaceholderText("Vorname*"), "Jane");
//     userEvent.type(screen.getByPlaceholderText("Nachname*"), "Doe");
//     userEvent.type(screen.getByPlaceholderText("E-Mail*"), "jane.doe@example.com");
//     userEvent.click(screen.getByRole("button", { name: /speichern/i }));
//     expect(jest.fn()).toHaveBeenCalled();
//     expect(jest.fn()).toHaveBeenCalled();
//   });

//   it("calls onUpdateUser when onSubmit is called with an existing contact", () => {
//     render(
//       <Form
//         contact={{ id: 1, firstname: "John", lastname: "Doe", email: "john.doe@example.com" }}
//         onAddUser={mockOnAddUser}
//         onUpdateUser={mockOnUpdateUser}
//         onDeleteUser={mockOnDeleteUser}
//         onCloseModal={mockOnCloseModal}
//       />,
//     );
//     userEvent.click(screen.getByRole("button", { name: /speichern/i }));
//     expect(jest.fn()).toHaveBeenCalled();
//     expect(jest.fn()).toHaveBeenCalled();
//   });

//   it("calls onDeleteUser when handleDelete is called", () => {
//     const contactWithId = {
//       id: 1,
//       firstname: "John",
//       lastname: "Doe",
//       email: "john.doe@example.com",
//     };
//     render(
//       <Form
//         contact={contactWithId}
//         onAddUser={mockOnAddUser}
//         onUpdateUser={mockOnUpdateUser}
//         onDeleteUser={mockOnDeleteUser}
//         onCloseModal={mockOnCloseModal}
//       />,
//     );
//     userEvent.click(screen.getByRole("button", { name: /l√∂schen/i }));
//     expect(mockOnDeleteUser).toHaveBeenCalledWith(contactWithId.id);
//     expect(mockOnCloseModal).toHaveBeenCalled();
//   });
// });
