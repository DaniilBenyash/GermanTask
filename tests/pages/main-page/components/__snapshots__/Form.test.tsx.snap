// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form
  data-testid="form"
  onSubmit={[Function]}
>
  <div
    className="data_container"
  >
    <h2
      className="title"
    >
      Eintrag bearbeiten
    </h2>
    <div
      className="input_container"
    >
      <div
        className="container"
      >
        <input
          className="input"
          data-testid="VornameInput"
          name="firstname"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Vorname*"
        />
        <label
          className="label"
        >
          Vorname*
        </label>
      </div>
      <div
        className="container"
      >
        <input
          className="input"
          data-testid="NachnameInput"
          name="lastname"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="Nachname*"
        />
        <label
          className="label"
        >
          Nachname*
        </label>
      </div>
      <div
        className="container"
      >
        <input
          className="input"
          data-testid="E-MailInput"
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="E-Mail*"
        />
        <label
          className="label"
        >
          E-Mail*
        </label>
      </div>
    </div>
  </div>
  <div
    className="button_container"
  >
    <div
      className="right_section_buttons"
    >
      <button
        className="button button_primary"
        onClick={[Function]}
      >
        ABBRECHEN
      </button>
      <button
        className="button button_default"
        disabled={true}
      >
        SPEICHERN
      </button>
    </div>
  </div>
</form>
`;
